File: Solutions/Python/word_search_ii.py
"""
    [HARD]
    212. Word Search II

    Concepts:
    - matrix
    - trie
    - depth-first search

    Stats:
        Runtime | 6488 ms   [Beats 32.23%]
        Memory  | 18.95 MB  [Beats 78.40%]
"""

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        """
            Build trie, then search trie and
            matrix using DFS
        """
        m, n = len(board), len(board[0])
        words = set(words)
        words_on_board = set()
        visited = set()
        moves = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        
        trie = {}
        for word in words:
            curr = trie
            for c in word:
                if c not in curr:
                    curr[c] = {}
                curr = curr[c]
            # end-of-word marker
            curr[None] = None
            
        def dfs(i: int, j: int, graph: dict, word: str) -> None:
            nonlocal words_on_board, visited
            
            # if whole word found...
            if None in graph:
                words_on_board.add(word)
            
            for di, dj in moves:
                ni, nj = i + di, j + dj
                if (
                    0 <= ni < m and
                    0 <= nj < n and
                    (ni, nj) not in visited and 
                    board[ni][nj] in graph
                ):
                    visited.add((i, j))
                    dfs(ni, nj, graph[board[ni][nj]], word + board[ni][nj])
                    visited.remove((i, j))
        
        for i in range(m):
            for j in range(n):
                if board[i][j] in trie:
                    visited.clear()
                    dfs(i, j, trie[board[i][j]], board[i][j])
                    
        return list(words_on_board)

File: Solutions/Cpp/word_search_ii.cpp
/*
    [HARD]
    212. Word Search II

    Concepts:
    - matrix
    - trie
    - depth-first search

    Stats:
        Runtime | 1179 ms   [Beats 20.73%]
        Memory  | 20.69 MB  [Beats 31.62%]
*/

class TrieNode {
public:
    std::unordered_map<char, TrieNode*> nodes;
    bool is_leaf, word_found;

    TrieNode() {
        nodes = {};
        is_leaf = false;
        word_found = false;
    }

    bool search(string word) {
        TrieNode* curr = this;

        for (const char c : word) {
            if (!curr->nodes.contains(c))
                return false;
            curr = curr->nodes[c];
        }
        return curr->is_leaf;
    }
};

class Solution {
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        /*
            Build trie, search trie and
            matrix using DFS
        */
        const int m = board.size(), n = board[0].size();
        const int moves[4][2] = {
            {0, 1},
            {0, -1},
            {1, 0},
            {-1, 0}
        };

        std::vector<std::string> words_on_board = {};
        TrieNode* root = new TrieNode();
        TrieNode* curr;

        // populate trie
        for (const std::string word : words) {
            curr = root;

            for (const char c : word) {
                if (!curr->nodes.contains(c))
                    curr->nodes[c] = new TrieNode();
                curr = curr->nodes[c];
            }
            curr->is_leaf = true;
        }

        std::vector<std::vector<bool>> visited(m, std::vector<bool>(n, false));
        std::function<void(int, int, std::string, TrieNode*)> dfs;

        dfs = [&](int i, int j, std::string path, TrieNode* curr){
            /*
                Performs DFS on board while using
                trie for guidance
            */
            if (curr->is_leaf && !curr->word_found) {
                words_on_board.push_back(path);
                curr->word_found = true;
            }

            for (const auto [di, dj] : moves) {
                int ni = i + di;
                int nj = j + dj;
                if (
                    0 <= ni && ni < m &&
                    0 <= nj && nj < n &&
                    !visited[ni][nj] &&
                    curr->nodes.contains(board[ni][nj])
                ) {
                    visited[ni][nj] = true;
                    dfs(ni, nj, path + board[ni][nj], curr->nodes[board[ni][nj]]);
                    visited[ni][nj] = false;
                }
            }
        };

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (root->nodes.contains(board[i][j])) {
                    visited = std::vector(m, std::vector<bool>(n, false));
                    visited[i][j] = true;
                    dfs(i, j, std::string{board[i][j]}, root->nodes[board[i][j]]);
                }
            }
        }

        return words_on_board;
    }
};